docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,255 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Redefine Container Runtime Interface
+
+The umbrella issue: [#22964](https://issues.k8s.io/22964)
+
+## Motivation
+
+Kubelet employs a declarative pod-level interface, which acts as the sole
+integration point for container runtimes (e.g., `docker` and `rkt`). The
+high-level, declarative interface has caused higher integration and maintenance
+cost, and also slowed down feature velocity for the following reasons.
+  1. **Not every container runtime supports the concept of pods natively**.
+     When integrating with Kubernetes, a significant amount of work needs to
+     go into implementing a shim of significant size to support all pod
+     features. This also adds maintenance overhead (e.g., `docker`).
+  2. **High-level interface discourages code sharing and reuse among runtimes**.
+     E.g, each runtime today implements an all-encompassing `SyncPod()`
+     function, with the Pod Spec as the input argument. The runtime implements
+     logic to determine how to achieve the desired state based on the current
+     status, (re-)starts pods/containers and manages lifecycle hooks
+     accordingly.
+  3. **Pod Spec is susceptible to inconsistent interpretations**. E.g., `rkt`
+     does not support container-level operations and assume immutable pods.
+  4. **Pod Spec is evolving rapidly**. New features are being added constantly.
```

Comment:
We don't have a well-defined pod spec because of the fact that pods are evolving. But, the new container runtime interface steers the pod concept towards a reserved set of resources in which containers can be spun. 

By allowing container-level operations, the unit of execution for k8s becomes containers instead of being application-oriented. This is very similar to Borg allocs, while they were useful, a better concept is the pod concept which is the smallest piece of the application that makes sense as a unit. 

A pod may be made up of many containers, but the state of the containers inside a pod must be atomic, whereas this proposal will support a finer-grained state mutations inside a pod.